pkg/skaffold/runner/runner.go

```diff
@@ -68,121 +64,129 @@ func NewForConfig(opts *config.SkaffoldOptions, cfg *config.SkaffoldConfig, out
 
 	deployer, err := getDeployer(&cfg.Deploy, kubeContext)
 	if err != nil {
 		return nil, errors.Wrap(err, "parsing skaffold deploy config")
 	}
+	deployer = deploy.WithTimings(deployer)
+	if opts.Notification {
+		deployer = deploy.WithNotification(deployer)
+	}
 
 	tagger, err := getTagger(cfg.Build.TagPolicy, opts.CustomTag)
 	if err != nil {
 		return nil, errors.Wrap(err, "parsing skaffold tag config")
 	}
 
-	client, err := kubernetesClient()
-	if err != nil {
-		return nil, errors.Wrap(err, "getting k8s client")
-	}
-
 	return &SkaffoldRunner{
-		config:         cfg,
-		Builder:        builder,
-		Deployer:       deployer,
-		Tagger:         tagger,
-		opts:           opts,
-		kubeclient:     client,
-		WatcherFactory: watch.NewWatcher,
-		out:            out,
+		config:               cfg,
+		Builder:              builder,
+		Deployer:             deployer,
+		Tagger:               tagger,
+		opts:                 opts,
+		WatcherFactory:       watch.NewWatcher,
+		DependencyMapFactory: build.NewDependencyMap,
+		out:                  out,
 	}, nil
 }
 
 func getBuilder(cfg *v1alpha2.BuildConfig, kubeContext string) (build.Builder, error) {
-	if cfg.LocalBuild != nil {
+	switch {
+	case cfg.LocalBuild != nil:
 		logrus.Debugf("Using builder: local")
 		return build.NewLocalBuilder(cfg, kubeContext)
-	}
-	if cfg.GoogleCloudBuild != nil {
+
+	case cfg.GoogleCloudBuild != nil:
 		logrus.Debugf("Using builder: google cloud")
 		return build.NewGoogleCloudBuilder(cfg)
-	}
-	if cfg.KanikoBuild != nil {
+
+	case cfg.KanikoBuild != nil:
 		logrus.Debugf("Using builder: kaniko")
 		return build.NewKanikoBuilder(cfg)
-	}
 
-	return nil, fmt.Errorf("Unknown builder for config %+v", cfg)
+	default:
+		return nil, fmt.Errorf("Unknown builder for config %+v", cfg)
+	}
 }
 
 func getDeployer(cfg *v1alpha2.DeployConfig, kubeContext string) (deploy.Deployer, error) {
-	if cfg.KubectlDeploy != nil {
+	switch {
+	case cfg.KubectlDeploy != nil:
 		return deploy.NewKubectlDeployer(cfg, kubeContext), nil
-	}
-	if cfg.HelmDeploy != nil {
+
+	case cfg.HelmDeploy != nil:
 		return deploy.NewHelmDeployer(cfg, kubeContext), nil
-	}
 
-	return nil, fmt.Errorf("Unknown deployer for config %+v", cfg)
+	default:
+		return nil, fmt.Errorf("Unknown deployer for config %+v", cfg)
+	}
 }
 
 func getTagger(t v1alpha2.TagPolicy, customTag string) (tag.Tagger, error) {
-	if customTag != "" {
+	switch {
+	case customTag != "":
 		return &tag.CustomTag{
 			Tag: customTag,
 		}, nil
-	}
 
-	if t.EnvTemplateTagger != nil {
+	case t.EnvTemplateTagger != nil:
 		return tag.NewEnvTemplateTagger(t.EnvTemplateTagger.Template)
-	}
-	if t.ShaTagger != nil {
+
+	case t.ShaTagger != nil:
 		return &tag.ChecksumTagger{}, nil
-	}
-	if t.GitTagger != nil {
+
+	case t.GitTagger != nil:
 		return &tag.GitCommit{}, nil
-	}
 
-	return nil, fmt.Errorf("Unknown tagger for strategy %s", t)
+	default:
+		return nil, fmt.Errorf("Unknown tagger for strategy %s", t)
```

Comment:
nit: `Unknown tagger for config %+v` since its a struct now

Reply:
Will do.